File: scala-3/file-share-system/src/main/scala/com/rbleggi/filesharesystem/CommandInvoker.scala
package com.rbleggi.filesharesystem

import scala.collection.mutable

class CommandInvoker {
  private val history = mutable.Stack[Command]()
  private val redoStack = mutable.Stack[Command]()

  def executeCommand(command: Command): Unit = {
    command.execute()
    history.push(command)
    redoStack.clear()
  }

  def undo(): Unit = {
    if (history.nonEmpty) {
      val command = history.pop()
      command.undo()
      redoStack.push(command)
    } else {
      println("No actions to undo.")
    }
  }

  def redo(): Unit = {
    if (redoStack.nonEmpty) {
      val command = redoStack.pop()
      command.execute()
      history.push(command)
    } else {
      println("No actions to redo.")
    }
  }
}

File: java/file-share/src/main/java/com/rbleggi/filesharesystem/CommandInvoker.java
package com.rbleggi.filesharesystem;

import java.util.Stack;

public class CommandInvoker {
    private final Stack<Command> history = new Stack<>();
    private final Stack<Command> redoStack = new Stack<>();

    public void executeCommand(Command command) {
        command.execute();
        history.push(command);
        redoStack.clear();
    }

    public void undo() {
        if (!history.isEmpty()) {
            Command command = history.pop();
            command.undo();
            redoStack.push(command);
        } else {
            System.out.println("No actions to undo.");
        }
    }

    public void redo() {
        if (!redoStack.isEmpty()) {
            Command command = redoStack.pop();
            command.execute();
            history.push(command);
        } else {
            System.out.println("No actions to redo.");
        }
    }
}

